crates/evm/src/fuzz/strategies/param.rs

```diff
@@ -1,149 +1,158 @@
+pub fn fuzz_param(param: &DynSolType) -> SBoxedStrategy<DynSolValue> {
     match param {
-        ParamType::Address => {
-            // The key to making this work is the `boxed()` call which type erases everything
-            // https://altsysrq.github.io/proptest-book/proptest/tutorial/transforming-strategies.html
-            any::<[u8; 20]>().prop_map(|x| H160(x).into_token()).boxed()
-        }
-        ParamType::Bytes => any::<Vec<u8>>().prop_map(|x| Bytes::from(x).into_token()).boxed(),
-        ParamType::Int(n) => {
-            super::IntStrategy::new(*n, vec![]).prop_map(|x| x.into_token()).boxed()
-        }
-        ParamType::Uint(n) => {
-            super::UintStrategy::new(*n, vec![]).prop_map(|x| x.into_token()).boxed()
-        }
-        ParamType::Bool => any::<bool>().prop_map(|x| x.into_token()).boxed(),
-        ParamType::String => any::<Vec<u8>>()
-            .prop_map(|x| Token::String(unsafe { String::from_utf8_unchecked(x) }))
-            .boxed(),
-        ParamType::Array(param) => proptest::collection::vec(fuzz_param(param), 0..MAX_ARRAY_LEN)
-            .prop_map(Token::Array)
-            .boxed(),
-        ParamType::FixedBytes(size) => {
-            prop::collection::vec(any::<u8>(), *size).prop_map(Token::FixedBytes).boxed()
+        DynSolType::Address |
+        DynSolType::Function |
+        DynSolType::Bool |
+        DynSolType::Uint(_) |
+        DynSolType::Int(_) |
+        DynSolType::Bytes => DynSolValue::type_strategy(param),
+        DynSolType::FixedBytes(size) => {
+            let size = *size;
+            DynSolValue::type_strategy(&DynSolType::FixedBytes(size))
         }
-        ParamType::FixedArray(param, size) => {
-            prop::collection::vec(fuzz_param(param), *size).prop_map(Token::FixedArray).boxed()
+        DynSolType::String => DynSolValue::type_strategy(param)
+            .prop_map(move |value| {
+                DynSolValue::String(
+                    String::from_utf8_lossy(value.as_str().unwrap().as_bytes())
+                        .trim()
+                        .trim_end_matches('\0')
+                        .to_string(),
+                )
+            })
+            .sboxed(),
+        DynSolType::Tuple(params) => {
+            params.iter().map(fuzz_param).collect::<Vec<_>>().prop_map(DynSolValue::Tuple).sboxed()
         }
-        ParamType::Tuple(params) => {
-            params.iter().map(fuzz_param).collect::<Vec<_>>().prop_map(Token::Tuple).boxed()
+        DynSolType::FixedArray(param, size) => {
+            let fixed_size = *size;
+            proptest::collection::vec(fuzz_param(param), fixed_size)
+                .prop_map(DynSolValue::FixedArray)
+                .sboxed()
         }
+        DynSolType::Array(param) => proptest::collection::vec(fuzz_param(param), 0..MAX_ARRAY_LEN)
+            .prop_map(DynSolValue::Array)
+            .sboxed(),
+        DynSolType::CustomStruct { .. } => panic!("unsupported type"),
     }
 }
```

Comment:
Wondering what's going on here—fuzzing failures in CI are due to this fn generating invalid values (e.g `uint256`s for `uint8`s and so on), making the function to fuzz revert instantly. Do these need to be constrained further? cc @DaniPopes